File: 合并两个有序链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2){
    if(l1==NULL)
        return l2;
    if(l2==NULL)
        return l1;
    if(l1->val<l2->val){
        l1->next=mergeTwoLists(l1->next,l2);
            return l1;
        }else{
        l2->next=mergeTwoLists(l1,l2->next);
            return l2;
    }
}

struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2){
	struct ListNode *p,q;
	if(l1==NULL)
		return l2;
	else if(l2==NULL)
		return l1;
	p=NULL;
	q=NULL;
	while(l1!=NULL&&l2!=NULL){
		if(l1->val<l2->val){
			if(p==NULL)
				p=q=l1;
			else{
				q->next=l1;
				q=q->next;
			}
			l1=l1->next;
		}
		else{
			if(p==NULL){
				p=q=l2;
			}
			else{
				q->next=l2;
				q=q->next;
			}
			l2=l2->next;
		}
}
		if(l1==NULL)
			q->next=l2;
		else
			q->next=l1;
		return p;
} 
 
 

File: 2020年11月/合并两个有序链表.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        res=ListNode(None) #合并之后的新数组
        p=res #保持头节点不动
        if l1==None:  
            res.next=l2
            return res.next
        if l2==None:
            res.next=l1
            return res.next
        while l1 and l2:
            if l1.val < l2.val:
                p.next=l1
                l1=l1.next
            else:
                p.next=l2
                l2=l2.next
            p=p.next
        if l1: #如果l1还没循环结束
            p.next=l1
        else: #如果l2还没循环结束
            p.next=l2
        return res.next

